exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go

```diff
@@ -19,65 +19,381 @@ package controllers
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups="",resources=secrets;events;configmaps;serviceaccounts,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=roles;rolebindings,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups="management.cattle.io",resources=*,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=rke2configs;rke2configs/status;rke2configs/finalizers,verbs=get;list;watch;create;update;patch;delete
 
-// Reconcile reconciles the EtcdSnapshotRestore object.
-func (r *EtcdSnapshotRestoreReconciler) Reconcile(_ context.Context, _ ctrl.Request) (ctrl.Result, error) {
+func (r *ETCDSnapshotRestoreReconciler) Reconcile(ctx context.Context, restore *snapshotrestorev1.ETCDSnapshotRestore) (_ ctrl.Result, reterr error) {
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(restore, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		// Patch ObservedGeneration only if the reconciliation completed successfully.
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+		}
+		if err := patchHelper.Patch(ctx, restore, patchOpts...); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Handle deleted etcdSnapshot
+	if !restore.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, restore)
+	}
+
+	var errs []error
+
+	result, err := r.reconcileNormal(ctx, restore)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error reconciling etcd snapshot restore: %w", err))
+	}
+
+	return result, kerrors.NewAggregate(errs)
+}
+
+func (r *ETCDSnapshotRestoreReconciler) reconcileNormal(ctx context.Context, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (_ ctrl.Result, reterr error) {
+	log := log.FromContext(ctx)
+
+	scope, err := initScope(ctx, r.Client, etcdSnapshotRestore)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	running := func(machine *clusterv1.Machine) bool {
+		return machine.Status.Phase == string(clusterv1.MachinePhaseRunning)
+	}
+
+	if scope.machines.Filter(running).Len() != scope.machines.Len() {
+		log.Info("Not all machines are running yet, requeuing")
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	if scope.machines.Filter(InitMachine(scope.etcdMachineSnapshot)).Len() != 1 {
+		return ctrl.Result{}, fmt.Errorf(
```

Comment:
does it make sense to mark restore as failed if machine doesn't exist?

Reply:
It does, however the machine may go away after the restore. This needs to return early if the restore was completed